src/BuildQueue.cpp

```diff
@@ -1252,35 +1260,96 @@ namespace t2
     if (queue->m_QuitSignalled)
       return false;
 
-    // If we're a worker thread, keep running until we quit.
-    if (0 != thread_index)
-      return true;
-
-    // We're the main thread. Just loop until there's no more nodes and then move on to the next pass.
-    return queue->m_PendingNodeCount > 0;
+    return true;
   }
-
+  
   static void BuildLoop(ThreadState* thread_state)
   {
     BuildQueue        *queue = thread_state->m_Queue;
     ConditionVariable *cv    = &queue->m_WorkAvailable;
     Mutex             *mutex = &queue->m_Lock;
-
+    Mutex             *maxJobsChangedMutex = &queue->m_MaxJobsChangedMutex;
+    ConditionVariable* maxJobsChangedConditionalVariable = &queue->m_MaxJobsChangedConditionalVariable;
     MutexLock(mutex);
+    bool waitingForWork = false;
+
+    auto HibernateForThrottlingIfRequired = [=]() {
+      //check if dynamic max jobs amount has been reduced to a point where we need this thread to hibernate.
+      //Don't take a mutex lock for this check, as this if check will almost never hit and it's in a perf critical loop.
+      if (thread_state->m_ThreadIndex < queue->m_DynamicMaxJobs)
+        return false;
+      
+      ProfilerScope profiler_scope("HibernateForThrottling", thread_state->m_ProfilerThreadId, nullptr, "thread_state_sleeping");
+
+      //Ok, it looks like we will need to hibernate. To do this, we need to first let go of the regular build mutex, as we
+      //are currently the thread that owns it.
+      MutexUnlock(mutex);
+
+      //now, we need to go and do a careful check again if we really need to hibernate.  Because we did the initial check without a lock,
+      //the max jobs value might have changed again by now. To do a safe check, we're going to take the max jobs change mutex lock here
```

Comment:
Does it actually matter if the max jobs value has changed since the previous check? What would happen if we didn't take this mutex and do this additional check?

Reply:
we would go and hibernate while we shouldn't, and we wouldn't be woken up, because the signal to do the wakeup could have been sent before we go to sleep